# 데이터 가공

## 패키지 임포트

In [ ]:
#넘파이와 판다스 라이브러리 가져오기
import numpy as np
import pandas as pd

In [210]:
# 데이터셋 불러오기
meteor_showers = pd.read_csv('C:\AI_School\PYTHONSTUDY\data\meteorshowers.csv')
moon_phases = pd.read_csv('C:\AI_School\PYTHONSTUDY\data\moonphases.csv')
constellations = pd.read_csv('data/constellations.csv')
cities = pd.read_csv('C:\AI_School\PYTHONSTUDY\data\cities.csv')

In [284]:
moon_phases = pd.read_csv('C:\AI_School\PYTHONSTUDY\data\moonphases.csv')

In [285]:
# 유성우 데이터 확인하기
meteor_showers.head()

name   radiant  bestmonth preferredhemisphere    enddate  startdate
0        Lyrids      Lyra          4            northern 2020-04-22 2020-04-21
1  Eta Aquarids  Aquarius          5            southern 2020-05-28 2020-04-19
2      Orionids     Orion         10  northern, southern 2020-11-07 2020-10-02
3      Perseids   Perseus          8            northern 2020-08-24 2020-07-14
4       Leonids       Leo         11  northern, southern 2020-11-30 2020-11-06

## 컬럼 dtype 변경(월 이름)

In [220]:
# 유성우 월 이름 변경
meteor_showers.head()

name   radiant bestmonth startmonth  startday  endmonth  endday  \
0        Lyrids      Lyra     april      april        21     april      22   
1  Eta Aquarids  Aquarius       may      april        19       may      28   
2      Orionids     Orion   october    october         2  november       7   
3      Perseids   Perseus    august       july        14    august      24   
4       Leonids       Leo  november   november         6  november      30   

           hemisphere preferredhemisphere  
0            northern            northern  
1  northern, southern            southern  
2  northern, southern  northern, southern  
3            northern            northern  
4  northern, southern  northern, southern

In [271]:
# 월명을 숫자로 변환하기 위한 딕셔너리 생성 months
months = {
    "january": 1, "february": 2, "march": 3, "april": 4,
    "may": 5, "june": 6, "july": 7, "august": 8,
    "september": 9, "october": 10, "november": 11, "december": 12
}

In [272]:
months

{'january': 1,
 'february': 2,
 'march': 3,
 'april': 4,
 'may': 5,
 'june': 6,
 'july': 7,
 'august': 8,
 'september': 9,
 'october': 10,
 'november': 11,
 'december': 12}

In [ ]:
#meteor_showers['bestmonth'].map(어떤식으로 매핑할것인지에 대한 함수또는 메소드 작성)
meteor_showers['bestmonth'].map(months) 

0     4
1     5
2    10
3     8
4    11
Name: bestmonth, dtype: int64

In [226]:
# 유성우데이터프레임의 월명을 숫자로 변환 map 메소드 사용(기존 apply사용도 가능)
# 각각의 besthmont, startmonth, endmonth를 숫자로 변환환다.

#meteor_showers['bestmonth'] = meteor_showers['bestmonth'].map(months)
#meteor_showers['startmonth'] = meteor_showers['startmonth'].map(months)
meteor_showers['endmonth'] = meteor_showers['endmonth'].map(months)


meteor_showers


name   radiant  bestmonth  startmonth  startday  endmonth  endday  \
0        Lyrids      Lyra          4           4        21         4      22   
1  Eta Aquarids  Aquarius          5           4        19         5      28   
2      Orionids     Orion         10          10         2        11       7   
3      Perseids   Perseus          8           7        14         8      24   
4       Leonids       Leo         11          11         6        11      30   

           hemisphere preferredhemisphere  
0            northern            northern  
1  northern, southern            southern  
2  northern, southern  northern, southern  
3            northern            northern  
4  northern, southern  northern, southern

In [ ]:
#한줄씩 실행하면 되는데 전체 실행하면 NaN으로 됨 → NaN으로 바뀌지않도록 해야함 : apply메소드사용해서 예외처리를해야함

def monthtonum(month):
    if isinstance(month, int) or pd.isna(month):
        return month
    return months.get(month, month)

meteor_showers['bestmonth'] = meteor_showers['bestmonth'].apply(monthtonum)

In [227]:
# 유성우 월명이 숫자로 변경되었는지 정보 확인
meteor_showers.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5 entries, 0 to 4
Data columns (total 9 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   name                 5 non-null      object
 1   radiant              5 non-null      object
 2   bestmonth            5 non-null      int64 
 3   startmonth           5 non-null      int64 
 4   startday             5 non-null      int64 
 5   endmonth             5 non-null      int64 
 6   endday               5 non-null      int64 
 7   hemisphere           5 non-null      object
 8   preferredhemisphere  5 non-null      object
dtypes: int64(5), object(4)
memory usage: 492.0+ bytes


In [228]:
# 다른 데이터프레임(달의위상, 별자리)도 월명을 숫자로 변환

In [288]:
moon_phases['month'] = moon_phases['month'].map(months)

In [274]:
# 달의위상
moon_phases

month  day      moonphase specialevent
0        1    1            NaN          NaN
1        1    2  first quarter          NaN
2        1    3            NaN          NaN
3        1    4            NaN          NaN
4        1    5            NaN          NaN
..     ...  ...            ...          ...
361     12   27            NaN          NaN
362     12   28            NaN          NaN
363     12   29      full moon          NaN
364     12   30            NaN          NaN
365     12   31            NaN          NaN

[366 rows x 4 columns]

In [233]:
constellations['bestmonth'] = constellations['bestmonth'].map(months)

In [235]:
# 별자리
constellations

constellation  bestmonth  latitudestart  latitudeend besttime hemisphere
0          Lyra          8             90          -40    21:00   northern
1      Aquarius         10             65          -90    21:00   southern
2         Orion          1             85          -75    21:00   northern
3       Perseus         12             90          -35    21:00   northern
4           Leo          4             90           65    21:00   northern

## 일자 정보 Datetime으로 변환(startdate 컬럼 생성)

In [236]:
# 4, 21 → 0421 → 20200421
2020 * 10000 + 4 * 100+ 21

20200421

In [237]:
2020 * 10000 + meteor_showers['startmonth']*100 + meteor_showers['startday']

0    20200421
1    20200419
2    20201002
3    20200714
4    20201106
dtype: int64

In [238]:
# 유성우 데이터프레임의 새로운 컬럼에 일자 정보를 datatime(20240428)으로 변환하여 저장, startdate, enddate
# startdate 컬럼 추가

pd.to_datetime(2020 * 10000 + meteor_showers['startmonth']*100 + meteor_showers['startday'],format='%Y%m%d')


0   2020-04-21
1   2020-04-19
2   2020-10-02
3   2020-07-14
4   2020-11-06
dtype: datetime64[ns]

In [240]:
# 유성우 데이터프레임의 새로운 컬럼에 일자 정보를 datatime(20240428)으로 변환하여 저장, startdate, enddate
# enddate 컬럼 추가

meteor_showers['startdate'] = pd.to_datetime(2020 * 10000 + meteor_showers['startmonth']*100 + meteor_showers['startday'],format='%Y%m%d')

#meteor_showers['enddate'] = pd.to_datetime(2020 * 10000 + meteor_showers['endmonth']*100 + meteor_showers['endday'],format='%Y%m%d')
meteor_showers

name   radiant  bestmonth  startmonth  startday  endmonth  endday  \
0        Lyrids      Lyra          4           4        21         4      22   
1  Eta Aquarids  Aquarius          5           4        19         5      28   
2      Orionids     Orion         10          10         2        11       7   
3      Perseids   Perseus          8           7        14         8      24   
4       Leonids       Leo         11          11         6        11      30   

           hemisphere preferredhemisphere    enddate  startdate  
0            northern            northern 2020-04-22 2020-04-21  
1  northern, southern            southern 2020-05-28 2020-04-19  
2  northern, southern  northern, southern 2020-11-07 2020-10-02  
3            northern            northern 2020-08-24 2020-07-14  
4  northern, southern  northern, southern 2020-11-30 2020-11-06

In [289]:
# moon_phases 데이터 프레임에 일자 정보를 datatime으로 변환하여 date컬럼 추가
moon_phases['date'] = pd.to_datetime(2020 * 10000 + moon_phases['month'] * 100 + moon_phases['day'],format='%Y%m%d')



In [276]:
moon_phases

month  day      moonphase specialevent       date
0        1    1            NaN          NaN 2020-01-01
1        1    2  first quarter          NaN 2020-01-02
2        1    3            NaN          NaN 2020-01-03
3        1    4            NaN          NaN 2020-01-04
4        1    5            NaN          NaN 2020-01-05
..     ...  ...            ...          ...        ...
361     12   27            NaN          NaN 2020-12-27
362     12   28            NaN          NaN 2020-12-28
363     12   29      full moon          NaN 2020-12-29
364     12   30            NaN          NaN 2020-12-30
365     12   31            NaN          NaN 2020-12-31

[366 rows x 5 columns]

## 달의 위상 데이터를 숫자로 변환

In [290]:
moon_phases['moonphase'].unique()

array([nan, 'first quarter', 'full moon', 'third quarter', 'new moon'],
      dtype=object)

In [291]:
# 달의 위상 데이터를 숫자로 변환하기
phases = {'new moon' : 0 , 'first quarter' : 0.5, 'third quarter' : 0.5 , 'full moon' : 1}
phases

{'new moon': 0, 'first quarter': 0.5, 'third quarter': 0.5, 'full moon': 1}

In [292]:
# 달의위상 데이터프레임에 phases데이터를 이용하여 백분율 컬럼(percentage) 새롭게 추가하기
#1.  
#moon_phases['moonphase'].map(phases) # dtype: float64 (실수) , NaN 있는 것도 확인됨

#2.위와 겹치지않도록 위에서 확인한 부분을 주석처리후 아래와 같이 처리
moon_phases['percentage'] = moon_phases['moonphase'].map(phases) 
moon_phases

#위까지하면 아직 NaN값은 남아있는것을 볼 수 있다. 

month  day      moonphase specialevent       date  percentage
0        1    1            NaN          NaN 2020-01-01         NaN
1        1    2  first quarter          NaN 2020-01-02         0.5
2        1    3            NaN          NaN 2020-01-03         NaN
3        1    4            NaN          NaN 2020-01-04         NaN
4        1    5            NaN          NaN 2020-01-05         NaN
..     ...  ...            ...          ...        ...         ...
361     12   27            NaN          NaN 2020-12-27         NaN
362     12   28            NaN          NaN 2020-12-28         NaN
363     12   29      full moon          NaN 2020-12-29         1.0
364     12   30            NaN          NaN 2020-12-30         NaN
365     12   31            NaN          NaN 2020-12-31         NaN

[366 rows x 6 columns]

## 불필요한 컬럼 삭제

In [24]:
# 각 데이터 프레임에서 불필요한 컬럼 삭제하기 


In [248]:
#meteor_showers 
#startmonth,startday, endomth,endday, hemisphere 삭제  --미리보기기
#meteor_showers.drop(columns=['startmonth','endmonth','endday','startday','hemisphere'])

#데이터프레임에 덮어씌우기
meteor_showers = meteor_showers.drop(columns=['startmonth','endmonth','endday','startday','hemisphere'])
print(meteor_showers.columns) 

Index(['name', 'radiant', 'bestmonth', 'preferredhemisphere', 'enddate',
       'startdate'],
      dtype='object')


In [294]:
moon_phases_backup = moon_phases.copy() #백업 만들기

#month,day, moonphase (percentage만들었음) 삭제 
#moon_phases.drop(columns=['month','day','moonphase'])


moon_phases = moon_phases.drop(columns=['month','day','moonphase'])
moon_phases


specialevent       date  percentage
0            NaN 2020-01-01         NaN
1            NaN 2020-01-02         0.5
2            NaN 2020-01-03         NaN
3            NaN 2020-01-04         NaN
4            NaN 2020-01-05         NaN
..           ...        ...         ...
361          NaN 2020-12-27         NaN
362          NaN 2020-12-28         NaN
363          NaN 2020-12-29         1.0
364          NaN 2020-12-30         NaN
365          NaN 2020-12-31         NaN

[366 rows x 3 columns]

In [250]:
#constellations_backup = constellations.copy() 백업 만들기
#constellations_backup = constellations.copy() 

#besttime 삭제
#constellations.drop(columns=['besttime'])

constellations = constellations.drop(columns=['besttime'])

constellations


constellation  bestmonth  latitudestart  latitudeend hemisphere
0          Lyra          8             90          -40   northern
1      Aquarius         10             65          -90   southern
2         Orion          1             85          -75   northern
3       Perseus         12             90          -35   northern
4           Leo          4             90           65   northern

## 달의 위상 누락값 처리

In [28]:
# 달의위상 데이터 프레임에서 percentage 컬럼의 결측치 처리하기
# 데이터 프레임의 각 행을 순회하면서 이전 위상값으로 채워넣기

# iterrows()를 사용하여 moon_phases 데이터프레임의 각 행을 순차적으로 반복하면서 행(row)과 그 행의 인덱스(index)를 가져옵니다.
# 결측값인 경우, lastphase에 저장된 값을 해당 행의 'percentage'에 채워 넣습니다. 즉, 이전에 확인된 마지막 'percentage' 값을 사용하여 결측값을 대체합니다.


In [295]:
moon_phases

specialevent       date  percentage
0            NaN 2020-01-01         NaN
1            NaN 2020-01-02         0.5
2            NaN 2020-01-03         NaN
3            NaN 2020-01-04         NaN
4            NaN 2020-01-05         NaN
..           ...        ...         ...
361          NaN 2020-12-27         NaN
362          NaN 2020-12-28         NaN
363          NaN 2020-12-29         1.0
364          NaN 2020-12-30         NaN
365          NaN 2020-12-31         NaN

[366 rows x 3 columns]

# 데이터 프레임의 각 행을 순회하면서 이전 위상값으로 채워넣기 를 생각하자. 
# 결측값인 경우, lastphase에 저장된 값을 해당 행의 'percentage'에 채워 넣습니다. 즉, 이전에 확인된 마지막 'percentage' 값을 사용하여 결측값을 대체합니다.
1행에서 0.5 2행에서 NaN 인데 , lastphase 변수에  0.5를 넣었다. (갱신됨) 왜냐면 기존에 0이 아니였기때문이다. 
2행이 NaN이어서 0.5를가짐 .....반복
363행에서 1.0 나왔으니까 1가져, 그뒤로는 NaN에 1.0 넣었음



In [ ]:

# iterrows()를 사용하여 moon_phases 데이터프레임의 각 행을 순차적으로 반복하면서 행(row)과 그 행의 인덱스(index)를 가져옵니다.
# 결측값인 경우 lastphase를  해당 행의 percentage에 채워넣는다. 
#lastphase = 0.0

#for i in moon_phases['percentage']:
    #print(i) --val이 i였음, 확인 후 val로 변경한거임. 
 #   val = moon_phases.loc[i,'percentage']
 #   if pd.isnull(val):
 #       moon_phases.loc[i,'percentage'] = lastphase
 #   else:
 #       lastphase = val
 #   print (i)


#아래는 itterows 를 사용
#lastphase = 0.0
#for index, row in moon_phases.iterrows():   
#    if pd.isna(row['percentage']): #pd.isnull(row['percentage']) 과 같다
#     moon_phases.at[index, 'percentage'] = lastphase # moon_phases.loc[index,'percentage'] = lastphase
#else : 
#    lastphase = row['percentage']


#ffill 을 사용하는 방법도 있다

In [ ]:
lastphase = 0.0

for index, row in moon_phases.iterrows():
    print(f"index: {index}, percentage: {row['percentage']}, lastphase: {lastphase}")  # 디버깅

    if pd.isna(row['percentage']):
        moon_phases.at[index, 'percentage'] = lastphase
    else:
        lastphase = row['percentage']


moon_phases

In [ ]:
print(moon_phases.loc[0:10, ['date', 'percentage']])


# 데이터 분석

## 위도 조회

In [ ]:
지정된 도시에서 각 유성우를 볼 수 있는 가능성이 높은 날짜를 예측하려고 합니다.

In [95]:
cities

city  latitude               country
0                  Abu Dhabi     24.47  United Arab Emirates
1                      Abuja      9.07               Nigeria
2                      Accra      5.55                 Ghana
3                  Adamstown    -25.07      Pitcairn Islands
4                Addis Ababa      9.02              Ethiopia
..                       ...       ...                   ...
251  Abidjan(former capital)      6.82           Ivory Coast
252                  Yaounde      3.87              Cameroon
253          Yaren(de facto)     -0.53                 Nauru
254                  Yerevan     40.18               Armenia
255                   Zagreb     45.82               Croatia

[256 rows x 3 columns]

In [98]:
# 도시 데이터프레임에서 검색한 열(검색한 도시의 위도)만 가져오기
latitude = cities.loc[cities['city'] == 'Abu Dhabi','latitude'] 
latitude

0    24.47
Name: latitude, dtype: float64

In [99]:
type(latitude)

pandas.core.series.Series

In [101]:
print(cities.loc[cities['city'] == 'Seoul','latitude'] )

202    37.55
Name: latitude, dtype: float64


### 도시명 입력 후 위도 반환

In [102]:
# 도시명을 입력하면 위도를 반환하는 함수 만들기

def predict_best_meteor_shower_viewing(city):
    latitude = cities.loc[credits['city'] == city, 'latitude'] .iloc[0]
    return latitude

In [36]:
print(predict_best_meteor_shower_viewing('Abu Dhabi'))

24.47


In [37]:
print(predict_best_meteor_shower_viewing('Seoul'))

37.55


## 특정 위도에서 볼 수 있는 별자리 조회

In [ ]:
constellations[(constellations['latitudestart'] >= latitude) & (constellations['latitudeend'] <= latitude)]['constellations'].tolist()

constellations

constellation  bestmonth  latitudestart  latitudeend hemisphere
0          Lyra          8             90          -40   northern
1      Aquarius         10             65          -90   southern
2         Orion          1             85          -75   northern
3       Perseus         12             90          -35   northern
4           Leo          4             90           65   northern

In [120]:
# 도시명을 입력하면 위도와 별자리 리스트를 반환하는 함수 만들기

def predict_best_meteor_shower_viewing(city):
    latitude = cities.loc[cities['city'] == city, 'latitude'] .iloc[0]
    constellation_list = constellations[(constellations['latitudestart'] >= latitude) & (constellations['latitudeend'] <= latitude)]['constellation'].tolist()

    return latitude, constellation_list


In [122]:
# 뉴질랜드 웰링턴으로 확인
predict_best_meteor_shower_viewing('Wellington')

(np.float64(-41.28), ['Aquarius', 'Orion'])

In [123]:
predict_best_meteor_shower_viewing('Seoul')

(np.float64(37.55), ['Lyra', 'Aquarius', 'Orion', 'Perseus'])

## 문자 안내 출력

In [322]:
cities

city  latitude               country
0                  Abu Dhabi     24.47  United Arab Emirates
1                      Abuja      9.07               Nigeria
2                      Accra      5.55                 Ghana
3                  Adamstown    -25.07      Pitcairn Islands
4                Addis Ababa      9.02              Ethiopia
..                       ...       ...                   ...
251  Abidjan(former capital)      6.82           Ivory Coast
252                  Yaounde      3.87              Cameroon
253          Yaren(de facto)     -0.53                 Nauru
254                  Yerevan     40.18               Armenia
255                   Zagreb     45.82               Croatia

[256 rows x 3 columns]

In [ ]:
cities.values # 도시명 전체 조회

In [130]:
city = '광주'


meteor_shower_string=''

if city not in cities.values: #cities.value에 city가 있는지?
        meteor_shower_string = '입력한 도시 '  +city + '에서는 유성우 서비스를 제공하고있지 않습니다.'
     

print(meteor_shower_string)



입력한 도시 광주에서는 유성우 서비스를 제공하고있지 않습니다.


In [131]:
# 도시명을 입력히면 위도와 별자리리스트와 메시지를 반환하는 함수

def predict_best_meteor_shower_viewing(city):
    
    #도시가 없을 때 처리 
    if city not in cities.values: #cities.value에 city가 있는지?
        meteor_shower_string = '입력한 도시 '  +city + '에서는 유성우 서비스를 제공하고있지 않습니다.'
        return meteor_shower_string
    
    latitude = cities.loc[cities['city'] == city, 'latitude'] .iloc[0]    
    constellation_list = constellations[(constellations['latitudestart'] >= latitude) & (constellations['latitudeend'] <= latitude)]['constellation'].tolist()


    #constellation_list가 비어이있을수 있으니 해당 처리가 필요함
    
    #consetllation_list가 존재하지않으면 , "city"에서는 볼 수 있는 별자리가 없습니다.
    if not constellation_list:
        meteor_shower_string= city+'에서 볼 수 있는 별자리가 없습니다.'

    #"city"에서는 " "별자리를 관측할 수 있습니다. 
    meteor_shower_string = f'{city}에서는 유성우 관측이 가능합니다.'

    
    return latitude, constellation_list, meteor_shower_string

In [132]:
wellington = predict_best_meteor_shower_viewing('Wellington')
wellington

(np.float64(-41.28), ['Aquarius', 'Orion'], 'Wellington에서는 유성우 관측이 가능합니다.')

In [133]:
predict_best_meteor_shower_viewing('광주')

'입력한 도시 광주에서는 유성우 서비스를 제공하고있지 않습니다.'

In [134]:
predict_best_meteor_shower_viewing('Bangkok')

(np.float64(13.75),
 ['Lyra', 'Aquarius', 'Orion', 'Perseus'],
 'Bangkok에서는 유성우 관측이 가능합니다.')

In [55]:
wellington

(np.float64(-41.28), ['Aquarius', 'Orion'], 'Wellington에서는 별자리 관측이 가능합니다.')

## 유성우 관측 기간 조회(선택한 도시에서 관측 가능한 별자리 추출)

In [52]:
meteor_showers.head()

name   radiant  bestmonth preferredhemisphere  startdate    enddate
0        Lyrids      Lyra          4            northern 2020-04-21 2020-04-22
1  Eta Aquarids  Aquarius          5            southern 2020-04-19 2020-05-28
2      Orionids     Orion         10  northern, southern 2020-10-02 2020-11-07
3      Perseids   Perseus          8            northern 2020-07-14 2020-08-24
4       Leonids       Leo         11  northern, southern 2020-11-06 2020-11-30

In [140]:
# 유성우 관측 시작일 가져오기
# 여기서는 Wellington의 결과값을 가지고 조회


In [136]:
meteor_showers_startdate = meteor_showers.loc[meteor_showers['radiant'] == 'Orion','startdate'].iloc[0]
meteor_showers_startdate

Timestamp('2020-10-02 00:00:00')

In [300]:
# 유성우 관측 종료일


In [301]:
meteor_showers_enddate = meteor_showers.loc[meteor_showers['radiant'] == 'Orion','enddate'].iloc[0]
meteor_showers_enddate

Timestamp('2020-11-07 00:00:00')

In [302]:
meteor_showers_startdate = meteor_showers.loc[meteor_showers['radiant'] == 'Orion','startdate'].iloc[0]
meteor_showers_enddate = meteor_showers.loc[meteor_showers['radiant'] == 'Orion','enddate'].iloc[0]

## 달의 위상 최저치 추출

In [ ]:
달 위상의 최소값(달에서 반사되는 빛의 최소량)을 찾을 수 있습니다. 이 예측 함수에서는 첫 번째 날짜만 확인합니다.

In [315]:
# 가장 어두운 날 찾기
moon_phase_list['percentage'].idxmin() #가장 백분율이 적은 행 인덱스 반환

np.int64(275)

In [317]:
# 가장 어두운 날 찾기
best_moon_date = moon_phase_list.loc[moon_phase_list['percentage'].idxmin()]['date']

best_moon_date

Timestamp('2020-10-02 00:00:00')

In [321]:
meteor_showers_startdate = meteor_showers.loc[meteor_showers['radiant'] == 'Orion', 'startdate'].iloc[0]
meteor_showers_enddate = meteor_showers[meteor_showers['radiant'] == 'Orion']['enddate'].iloc[0]

moon_phase_list = moon_phases.loc[(moon_phases['date'] >= meteor_showers_startdate) & (moon_phases['date'] <= meteor_showers_enddate)]
moon_phase_list

best_moon_date = moon_phase_list.loc[moon_phase_list['percentage'].idxmin()]['date']
best_moon_date

Timestamp('2020-10-16 00:00:00')

## 최적의 유성우 관측일

In [324]:
# 도시명을 입력하면 관측가능한 유성우에 대한 정보를 알려주는 함수 작성하기

def predict_best_meteor_shower_viewing(city):
    # 안내 메시지 초기화
    meteor_shower_string = ''

    # 입력한 도시 정보가 존재하지 않으면 오류 메시지 그리고 return 으로 함수를 종료합니다.
    if city not in cities.values:
        meteor_shower_string = '입력하신'+city+'에서는 유성우 관측이 불가합니다.'
        return meteor_shower_string  
    
    # 도시의 위도 정보를 불러오기
    
    # 도시의 이름이 주어졌을때, 해당 도시의 위도(latitude) 값 이 필요하다는 의미이다. 
    #1.입력한 도시의 이름이 cities(데이터프레임)안에 존재하는지 확인하기 위해, 도시이름이 같은 지 비교하는 조건문을 작성한다. 
    # cities['city'] == city  
    #2. 위의 조건문을 작성했으니 이제 조건을 만족하는 행에서 필요한 컬럼을 뽑아와야한다.
    # .loc[조건, 컬럼명]
    # ▶ cities 데이터 프레임에서, city 컬럼에  입력한 'city'가 있는 행 중 latitude 컬럼만 뽑아옵니다. 
    # cities.loc[cities['city'] == city,'latitude']
    #3. iloc[0] : 위에서 뽑아낸 결과값에서 float 숫자 하나만 필요하기 때문에 시리즈의 첫번째 값을 꺼내옵니다. → Series 타입은 리스트 처럼 생긴 데이터를 의미하는 객체인데,  series 란 한 열 또는 하나의 값 목록입니다. 인덱스가 있어서 리스트 보다 유연하다. 
    latitude = cities.loc[cities['city'] == city,'latitude'].iloc[0]
    

    # 해당 도시의 위도에서 관측 가능한 별자리 조회
    constellation_list = constellations[(constellations['latitudestart'] >= latitude) & (constellations['latitudeend'] <= latitude)]['constellation'].tolist()



    # 볼 수 있는 별자리가 존재하지 않으면 오류 메시지
    if not constellation_list:
        meteor_shower_string = f"아쉽지만 {city}에서는 볼 수 있는 별자리가 없습니다."
        return meteor_shower_string

    
    # 별자리 관측 가능 안내 메시지
    meteor_shower_string = f'{city}에서는 유성우 관측이 가능합니다.\n'

    
    # 도시별 관찰 가능한 별자리 iterate
    for constellation in constellation_list:
        # 별자리와 가장 가까운 유성우 조회
        meteor_shower = meteor_showers.loc[meteor_showers['radiant'] == constellation, 'name'].iloc[0]


        # 유성우 관측이 가능한 시작일과 종료일
        meteor_showers_startdate = meteor_showers.loc[meteor_showers['radiant'] == constellation, 'startdate'].iloc[0]


        # 유성우가 보일 때의 달의 위상 조회
        meteor_showers_enddate = meteor_showers.loc[meteor_showers['radiant'] == constellation,'enddate'].iloc[0]

        # 달이 눈에 보이는 첫 날 조회
        best_moon_date = moon_phase_list.loc[moon_phase_list['percentage'].idxmin()]['date']



        # 사용자에게 정보전달
        meteor_shower_string += f"{meteor_shower}를 관측하려면 {constellation}자리 위치를 향해서 {best_moon_date.to_pydatetime().strftime('%Y년 %m월 %d일')}에 볼 수 있습니다.\n"



    return meteor_shower_string

In [325]:
print(predict_best_meteor_shower_viewing('New York'))

입력하신New York에서는 유성우 관측이 불가합니다.


In [326]:
print(predict_best_meteor_shower_viewing('Wellington'))

Wellington에서는 유성우 관측이 가능합니다.
Eta Aquarids를 관측하려면 Aquarius자리 위치를 향해서 2020년 10월 16일에 볼 수 있습니다.
Orionids를 관측하려면 Orion자리 위치를 향해서 2020년 10월 16일에 볼 수 있습니다.



In [327]:
print(predict_best_meteor_shower_viewing('Seoul'))

Seoul에서는 유성우 관측이 가능합니다.
Lyrids를 관측하려면 Lyra자리 위치를 향해서 2020년 10월 16일에 볼 수 있습니다.
Eta Aquarids를 관측하려면 Aquarius자리 위치를 향해서 2020년 10월 16일에 볼 수 있습니다.
Orionids를 관측하려면 Orion자리 위치를 향해서 2020년 10월 16일에 볼 수 있습니다.
Perseids를 관측하려면 Perseus자리 위치를 향해서 2020년 10월 16일에 볼 수 있습니다.



# 항아 유성우 데이터 추가

In [60]:
# 항아 유성우 데이터 추가하기
change_meteor_shower = {'name' : "Chang\'e", 'radiant' : 'Draco', 'bestmonth' : 'october', 'startmonth': 'october', 'startday' : 1,
                        'endmonth' : 'october', 'endday' : 31, 'hemisphere' : 'northern', 'preferredhemisphere' : 'northern' }
change_meteor_shower_df = pd.DataFrame(change_meteor_shower,index = [0])
change_meteor_shower_df

name radiant bestmonth startmonth  startday endmonth  endday hemisphere  \
0  Chang'e   Draco   october    october         1  october      31   northern   

  preferredhemisphere  
0            northern

In [61]:
# meteorshowers.csv 파일 읽어와서 new_meteor_showers 변수에 저장
new_meteor_showers = pd.read_csv('meteorshowers.csv')
new_meteor_showers

name   radiant bestmonth startmonth  startday  endmonth  endday  \
0        Lyrids      Lyra     april      april        21     april      22   
1  Eta Aquarids  Aquarius       may      april        19       may      28   
2      Orionids     Orion   october    october         2  november       7   
3      Perseids   Perseus    august       july        14    august      24   
4       Leonids       Leo  november   november         6  november      30   

           hemisphere preferredhemisphere  
0            northern            northern  
1  northern, southern            southern  
2  northern, southern  northern, southern  
3            northern            northern  
4  northern, southern  northern, southern

In [ ]:
# new_meteor_showers 데이터프레임에 change_meteor_shower_df 데이터프레임 추가하기 concat
new_meteor_showers = # TODO
new_meteor_showers

name   radiant bestmonth startmonth  startday  endmonth  endday  \
0        Lyrids      Lyra     april      april        21     april      22   
1  Eta Aquarids  Aquarius       may      april        19       may      28   
2      Orionids     Orion   october    october         2  november       7   
3      Perseids   Perseus    august       july        14    august      24   
4       Leonids       Leo  november   november         6  november      30   
5       Chang'e     Draco   october    october         1   october      31   

           hemisphere preferredhemisphere  
0            northern            northern  
1  northern, southern            southern  
2  northern, southern  northern, southern  
3            northern            northern  
4  northern, southern  northern, southern  
5            northern            northern

In [63]:
# Chang'e 유성우에 대해 Draconids 유성우가 10월 초에 발생할 가능성이 높은 용자리를 선택합니다. 이 유성우를 가상 유성우의 참조로 사용
# 용자리 항목 추가하기기
draco_constellation = {'constellation' : 'Draco', 'bestmonth' : 'july', 'latitudestart' : 90, 'latitudeend' : -15, 'besttime' :2100, 'hemisphere' : 'northern'}
draco_constellation_df = pd.DataFrame(draco_constellation, index = [0])
draco_constellation_df

constellation bestmonth  latitudestart  latitudeend  besttime hemisphere
0         Draco      july             90          -15      2100   northern

In [64]:
# 별자리 파일 읽어와서 new_constellations에 저장하기
new_constellations = pd.read_csv('constellations.csv')
new_constellations

constellation bestmonth  latitudestart  latitudeend besttime hemisphere
0          Lyra    august             90          -40    21:00   northern
1      Aquarius   october             65          -90    21:00   southern
2         Orion   january             85          -75    21:00   northern
3       Perseus  december             90          -35    21:00   northern
4           Leo     april             90           65    21:00   northern

In [ ]:
# new_constellations 데이터프레임에 draco_constellation_df 추가하기 concat
new_constellations =  # TODO
new_constellations

constellation bestmonth  latitudestart  latitudeend besttime hemisphere
0          Lyra    august             90          -40    21:00   northern
1      Aquarius   october             65          -90    21:00   southern
2         Orion   january             85          -75    21:00   northern
3       Perseus  december             90          -35    21:00   northern
4           Leo     april             90           65    21:00   northern
5         Draco      july             90          -15     2100   northern

In [123]:
# new_meteor_showers 데이터프레임 월명 매핑


name   radiant  bestmonth  startmonth  startday  endmonth  endday  \
0        Lyrids      Lyra          4           4        21         4      22   
1  Eta Aquarids  Aquarius          5           4        19         5      28   
2      Orionids     Orion         10          10         2        11       7   
3      Perseids   Perseus          8           7        14         8      24   
4       Leonids       Leo         11          11         6        11      30   
5       Chang'e     Draco         10          10         1        10      31   

           hemisphere preferredhemisphere  
0            northern            northern  
1  northern, southern            southern  
2  northern, southern  northern, southern  
3            northern            northern  
4  northern, southern  northern, southern  
5            northern            northern

In [124]:
# new_constellations 데이터프레임 월명 매핑

new_constellations

constellation  bestmonth  latitudestart  latitudeend besttime hemisphere
0          Lyra          8             90          -40    21:00   northern
1      Aquarius         10             65          -90    21:00   southern
2         Orion          1             85          -75    21:00   northern
3       Perseus         12             90          -35    21:00   northern
4           Leo          4             90           65    21:00   northern
5         Draco          7             90          -15     2100   northern

In [125]:
# new_meteor_showers 날짜형식 바꿔서 새로운 컬럼으로 추가 startdate, enddate


new_meteor_showers

name   radiant  bestmonth  startmonth  startday  endmonth  endday  \
0        Lyrids      Lyra          4           4        21         4      22   
1  Eta Aquarids  Aquarius          5           4        19         5      28   
2      Orionids     Orion         10          10         2        11       7   
3      Perseids   Perseus          8           7        14         8      24   
4       Leonids       Leo         11          11         6        11      30   
5       Chang'e     Draco         10          10         1        10      31   

           hemisphere preferredhemisphere  startdate    enddate  
0            northern            northern 2020-04-21 2020-04-22  
1  northern, southern            southern 2020-04-19 2020-05-28  
2  northern, southern  northern, southern 2020-10-02 2020-11-07  
3            northern            northern 2020-07-14 2020-08-24  
4  northern, southern  northern, southern 2020-11-06 2020-11-30  
5            northern            northern 2020-10-01 2020-10-31

In [126]:
# 도시명을 입력히면 별자리 정보를 알려주는 함수 완성하기
def predict_best_meteor_shower_viewing(city):
    


    return meteor_shower_string

In [127]:
print(predict_best_meteor_shower_viewing('Beijing'))

Beijing에서는 별자리 관측이 가능합니다: 

Lyrids를 잘 보려면 Lyra자리 위치를 향하여 April 22, 2020보면 됩니다.
Eta Aquarids를 잘 보려면 Aquarius자리 위치를 향하여 April 22, 2020보면 됩니다.
Orionids를 잘 보려면 Orion자리 위치를 향하여 October 16, 2020보면 됩니다.
Perseids를 잘 보려면 Perseus자리 위치를 향하여 July 20, 2020보면 됩니다.
Chang'e를 잘 보려면 Draco자리 위치를 향하여 October 16, 2020보면 됩니다.

